### PR TITLE
Builds on #27049 to extract getPriceSetID()

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Builds on https://github.com/civicrm/civicrm-core/pull/27049 to extract `getPriceSetID()`

Before
----------------------------------------
`discountID` (newly) pass in to `CRM_Event_Form_Registration::initEventFee(` - but used only to get `priceSetID`

After
------------------------------------------
`$priceSetID` passed in

Note that the `getPriceSetID` function now handles setting the form property & form value, as used. So I removed that from the shared code. I didn't put it in the setting for `ParticipantFeeSelection` because it is currently interacting with an undefined property

Technical Details
-----------------------------------------

Per https://github.com/civicrm/civicrm-core/pull/27049 the only other usage is a unit test @mlutfy wrote

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
test this please

File: CRM/Event/Form/ParticipantFeeSelection.php [hunk 1]
```diff
@@ -426,4 +426,30 @@ public function getEventID(): int {
...
+   * @return int|null
+   *
+   * @api This function will not change in a minor release and is supported for
+   * use outside of core. This annotation / external support for properties
+   * is only given where there is specific test cover.
+   *
+   * @noinspection PhpUnhandledExceptionInspection
+   * @noinspection PhpDocMissingThrowsInspection
+   */
+  public function getPriceSetID(): ?int {
```
Comment (larssandergreen) [inline CRM/Event/Form/ParticipantFeeSelection.php:441]: Should probably use this at 127 as well for consistency.

Comment (larssandergreen) [reply CRM/Event/Form/ParticipantFeeSelection.php:441]: Actually, $priceSetId is unused, we can just remove 127.

File: CRM/Event/Form/Participant.php [hunk 2]
```diff
@@ -1428,7 +1428,7 @@ public function buildEventFeeForm($form) {
 
       //retrieve custom information
       $form->_values = [];
-      CRM_Event_Form_Registration::initEventFee($form, $event['id'], FALSE, $form->getDiscountID());
+      CRM_Event_Form_Registration::initEventFee($form, $event['id'], FALSE, $this->getPriceSetID());
```
Comment (larssandergreen) [inline CRM/Event/Form/Participant.php:1431]: I don't think it makes any difference as things stand now, but shouldn't it be $form->getPriceSetID for consistency?

Comment (eileenmcnaughton) [reply CRM/Event/Form/Participant.php:1431]: Hmm - I've been trying to go the other way for consistency - ditch the use of form & even the passing of it

File: CRM/Event/Form/ParticipantFeeSelection.php (hunk 1 above)
Comment (eileenmcnaughton) [reply CRM/Event/Form/ParticipantFeeSelection.php:440]: done!

File: CRM/Event/Form/Participant.php (hunk 2 above)
Comment (larssandergreen) [reply CRM/Event/Form/Participant.php:1431]: Agreed. I guess it doesn't matter here as long as this doesn't get called from anywhere else, just looks a bit odd.

Comment (seamuslee001) [reply CRM/Event/Form/Participant.php:1431]: Not sure if this is affected or not https://github.com/eileenmcnaughton/civicrm-core/blob/master/tests/phpunit/CRM/Event/Form/ParticipantTest.php#L360

Comment (eileenmcnaughton) [reply CRM/Event/Form/Participant.php:1431]: @seamuslee001 yeah that needs cleaning up - although as long as this PR passes tests it is out of scope for it

**Comment by larssandergreen** [comment]
Also noting, though I suppose it is technically out of scope here as it is the current behaviour on master: if I change selections for a registration with a discounted price set, I expect the options to show the discounted price set, but what I get is the current price set based on the current date. For example, if someone registered for a student ticket with an early bird discount and I need to change them to an adult ticket, I would expect they would still get the early bird price.

I thought initEventFee was supposed to get the price set that the participant currently had based on the discount id, but that doesn't seem to be happening.

**Comment by eileenmcnaughton** [comment]
Oh - the issue is that `ParticipantFeeForm` no longer has the undefined property set `_priceSetId` - ideally we would pass it into the function rather than require it from the form - just trying to see where it is called from / when it would NOT be set....

image

File: CRM/Event/Form/Registration/Register.php [hunk 3]
```diff
@@ -423,7 +423,7 @@ public function buildQuickForm() {
       // build amount only when needed, skip incase of event full and waitlisting is enabled
       // and few other conditions check preProcess()
       if (!$this->_noFees) {
-        self::buildAmount($this);
+        self::buildAmount($this, FALSE, NULL, $this->_priceSetId);
```
Comment (larssandergreen) [inline CRM/Event/Form/Registration/Register.php:426]: $required was TRUE by default before, now FALSE - intentional?

**Comment by larssandergreen** [comment]
Confirming that missing price set selections is fixed.

The discounted price set still isn't loading when editing a participant. I noticed that it looks like the discount id isn't loading when editing the registration at all. I think we should see a discount set here when the registration used a discounted price set.
image

File: CRM/Event/Form/Registration/Register.php (hunk 3 above)
Comment (eileenmcnaughton) [reply CRM/Event/Form/Registration/Register.php:426]: opps  no

**Comment by eileenmcnaughton** [comment]
whoa that discount code is a can of worms - it adds the property to the form only if the property context exists - which is only true for one of the forms involved `Participant` - but that ONLY uses if within the same function. I've moved that back to the form as a first step - it might be good to fix the ParticipantFeeSelection not offering the discount in a follow up as this should be a refactor only. I guess I need to check if only back-office users can use that form cos only back office users should get the choice.

Hmm that's a question - it makes sense that a back-office user can change the event selection within the same time-based discount  set- but should a self-service user? I suspect not - if they want to do that they maybe have to contact  the organization as the discount might depend on them paying on time & they could work the system by buying a cheap ticket early & getting the expensive one later

**Comment by larssandergreen** [comment]
OK, makes sense. Yes, let's do the discount thing separately.

Looks like the test fails are related.

**Comment by eileenmcnaughton** [comment]
Yeah - but super odd 

[Test Result](https://test.civicrm.org/job/CiviCRM-Core-Matrix-PR/4034/BKPROF=dfl,SUITES=phpunit-api4,label=bknix-tmp/testReport/) (2 failures / +2)
[api\v4\Entity\ConformanceTest.testEntitiesProvider](https://test.civicrm.org/job/CiviCRM-Core-Matrix-PR/4034/BKPROF=dfl,SUITES=phpunit-api4,label=bknix-tmp/testReport/junit/(root)/api_v4_Entity_ConformanceTest/testEntitiesProvider/)
[api\v4\Entity\ConformanceTest.testConformance with data set "Discount"](https://test.civicrm.org/job/CiviCRM-Core-Matrix-PR/4034/BKPROF=dfl,SUITES=phpunit-api4,label=bknix-tmp/testReport/junit/(root)/api_v4_Entity_ConformanceTest/testConformance_with_data_set__Discount_/)

**Comment by eileenmcnaughton** [comment]
Ah - https://github.com/civicrm/civicrm-core/pull/27050/files#diff-dd32c0ff4ad7da7103eec712b4eadc292638dd9409343725d532ffc565fc7bb1

**Comment by larssandergreen** [comment]
Now just a concurrency test fail, so good.

File: CRM/Event/Form/Registration/Register.php [hunk 4]
```diff
@@ -540,12 +540,13 @@ public function buildQuickForm() {
    *   Form object.
    * @param bool $required
    *   True if you want to add formRule.
-   * @param int $discountId
+   * @param null $discountId
```
Comment (larssandergreen) [inline CRM/Event/Form/Registration/Register.php:543]: int|null

**Comment by larssandergreen** [comment]
Another [small related bug in master.](https://lab.civicrm.org/dev/core/-/issues/4526) Not an issue here.

**Comment by larssandergreen** [comment]
OK, I've given this a fairly thorough run and I think it is ready to merge.

**Comment by larssandergreen** [comment]
Excepting that one docblock issue above

**Comment by eileenmcnaughton** [comment]
I fixed the docblock & the test magically passed

**Pull request merged**